File: ts/test/chat/privacy/mongo/deletion.ts
import { ObjectId, UpdateFilter } from "mongodb";
import { MongoLocator } from "../../../../src";
import GroupChat from "../../model/gc";
import Message from "../../model/message";
import { FirestoreCollections } from "../../model/shared";
import User from "../../model/user";
import DirectMessage from "../../model/dm";

export default function handleDeletionMongo(dataSubjectId: string, locator: MongoLocator, obj: any): {
    nodesToTraverse: MongoLocator[],
    deleteNode: boolean,
    fieldsToUpdate?: UpdateFilter<any> | Partial<any>
} {
    switch (locator.dataType) {
        case 'user':
            const user = obj as User;
            user.id = obj._id.toString();
            return handleDeletionUser(dataSubjectId, locator, user);
        case 'groupChat':
            const groupChat = obj as GroupChat;
            groupChat.id = obj._id.toString();
            return handleDeletionGroupChat(dataSubjectId, locator, groupChat);
        case 'directMessage':
            const directMessage = obj as DirectMessage;
            directMessage.id = obj._id.toString();
            return handleDeletionDirectMessage(dataSubjectId, locator, directMessage);
        case 'message':
            const message = obj as Message;
            message.id = obj._id.toString();
            return handleDeletionMessage(dataSubjectId, locator, message);
        default:
            throw new Error(`Data type ${locator.dataType} does not have handleDeletion function implemented`);
    }
}

function handleDeletionGroupChat(dataSubjectId: string, locator: MongoLocator, obj: GroupChat): {
    nodesToTraverse: MongoLocator[],
    deleteNode: boolean,
    fieldsToUpdate?: UpdateFilter<any> | Partial<any>
} {
    return {
        nodesToTraverse: [{
            dataType: 'message',
            singleDocument: false,
            collection: FirestoreCollections.Messages,
            filter: {
                userID: dataSubjectId,
                chatID: obj.id
            },
            context: {
                anonymize: true
            }
        }],
        deleteNode: obj.owner === dataSubjectId,
        fieldsToUpdate: obj.owner === dataSubjectId ? undefined :
            // remove dataSubjectId from users field
            {
                $pull: {
                    users: dataSubjectId
                }
            }
    };
};

function handleDeletionUser(dataSubjectId: string, locator: MongoLocator, obj: User): {
    nodesToTraverse: MongoLocator[],
    deleteNode: boolean,
    fieldsToUpdate?: UpdateFilter<any> | Partial<any>
} {
    if (obj.id !== dataSubjectId) {
        throw new Error(`User ${dataSubjectId} cannot delete another user ${obj.id}`);
    }

    const GCsToTraverse = obj.gcs.map((gc: string): MongoLocator => {
        return {
            dataType: 'groupChat',
            singleDocument: true,
            collection: FirestoreCollections.GroupChat,
            filter: {
                _id: new ObjectId(gc)
            }
        }
    })

    const DMsToTraverse = Object.values(obj.dms).map((dm: string): MongoLocator => {
        return {
            dataType: 'directMessage',
            singleDocument: true,
            collection: FirestoreCollections.DirectMessages,
            filter: {
                _id: new ObjectId(dm)
            }
        }
    })

    return {
        nodesToTraverse: [...GCsToTraverse, ...DMsToTraverse],
        deleteNode: true,
    };
}

function handleDeletionDirectMessage(dataSubjectId: string, locator: MongoLocator, obj: DirectMessage): {
    nodesToTraverse: MongoLocator[],
    deleteNode: boolean,
    fieldsToUpdate?: UpdateFilter<any> | Partial<any>
} {
    let otherUserId: string;
    let thisUserField: string;
    if (obj.user1 === dataSubjectId) {
        thisUserField = 'user1';
        otherUserId = obj.user2;
    } else if (obj.user2 === dataSubjectId) {
        thisUserField = 'user2';
        otherUserId = obj.user1;
    } else {
        throw new Error(`User ${dataSubjectId} does not have access to direct message ${obj.id}`);
    }

    return {
        // TODO: locator should probably have some sort of context to be passed to the next object
        nodesToTraverse: [{
            dataType: 'message',
            singleDocument: false,
            collection: FirestoreCollections.Messages,
            filter: {
                userID: dataSubjectId,
                chatID: obj.id
            },
            context: {
                anonymize: true
            }
        }],
        deleteNode: otherUserId === "anonymous",
        fieldsToUpdate: {
            $set: {
                [thisUserField]: 'anonymous'
            }
        }
    };
}

function handleDeletionMessage(dataSubjectId: string, locator: MongoLocator, obj: Message): {
    nodesToTraverse: MongoLocator[],
    deleteNode: boolean,
    fieldsToUpdate?: UpdateFilter<any> | Partial<any>
} {
    const fieldsToUpdate = (obj.userID === dataSubjectId && locator.context.anonymize) ?
        {
            $set: {
                userID: 'anonymous'
            }
        } : undefined;
    return {
        nodesToTraverse: [],
        deleteNode: obj.userID === dataSubjectId && !locator.context.anonymize,
        fieldsToUpdate: fieldsToUpdate
    };
};

File: go/internal/test/chat/deletion.go
package chat

import (
	"log"

	"cloud.google.com/go/firestore"
	pal "github.com/privacy-pal/privacy-pal/go/pkg"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func HandleDeletion(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) (nodesToTraverse []pal.Locator, deleteNode bool, fieldsToUpdate pal.FieldUpdates) {
	switch currentDbObjLocator.DataType {
	case UserDataType:
		return handleDeletionUser(dataSubjectId, currentDbObjLocator, dbObj)
	case GroupChatDataType:
		return handleDeletionGroupChat(dataSubjectId, currentDbObjLocator, dbObj)
	case MessageDataType:
		return handleDeletionMessage(dataSubjectId, currentDbObjLocator, dbObj)
	case DirectMessageDataType:
		return handleDeletionDirectMessage(dataSubjectId, currentDbObjLocator, dbObj)
	default:
		// TODO: should return error
		return nil, false, pal.FieldUpdates{}
	}
}

func handleDeletionUser(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) (nodesToTraverse []pal.Locator, deleteNode bool, fieldsToUpdate pal.FieldUpdates) {
	for _, id := range dbObj["gcs"].([]interface{}) {
		id := id.(string)
		objectID, err := primitive.ObjectIDFromHex(id)
		if err != nil {
			log.Println(err)
		}

		nodesToTraverse = append(nodesToTraverse, pal.Locator{
			LocatorType: pal.Document,
			DataType:    GroupChatDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"gcs"},
				DocIDs:         []string{id},
			},
			MongoLocator: pal.MongoLocator{
				Collection: "gcs",
				Filter:     bson.D{{Key: "_id", Value: objectID}},
			},
		})
	}

	for _, dmID := range dbObj["dms"].(map[string]interface{}) {
		dmID := dmID.(string)
		dmIDObj, err := primitive.ObjectIDFromHex(dmID)
		if err != nil {
			log.Println(err)
		}

		nodesToTraverse = append(nodesToTraverse, pal.Locator{
			LocatorType: pal.Document,
			DataType:    DirectMessageDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"dms"},
				DocIDs:         []string{dmID},
			},
			MongoLocator: pal.MongoLocator{
				Collection: "dms",
				Filter:     bson.D{{Key: "_id", Value: dmIDObj}},
			},
		})
	}

	deleteNode = true
	return
}

func handleDeletionGroupChat(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) (nodesToTraverse []pal.Locator, deleteNode bool, fieldsToUpdate pal.FieldUpdates) {
	nodesToTraverse = append(nodesToTraverse, pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Filters: []pal.Filter{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
		MongoLocator: pal.MongoLocator{
			Collection: "messages",
			Filter:     bson.D{{Key: "userId", Value: dataSubjectId}, {Key: "chatId", Value: dbObj["_id"]}},
		},
	})

	deleteNode = false

	firestoreUpdates := []firestore.Update{}
	mongoUpdates := []interface{}{}

	if (dbObj["owner"] == dataSubjectId) && (len(dbObj["users"].([]interface{})) > 1) {
		firestoreUpdates = append(firestoreUpdates, firestore.Update{
			Path:  "owner",
			Value: dbObj["users"].([]interface{})[0],
		})
		mongoUpdates = append(mongoUpdates, bson.D{{Key: "owner", Value: dbObj["users"].([]interface{})[0]}})
	} else {
		firestoreUpdates = append(firestoreUpdates, firestore.Update{
			Path:  "users",
			Value: firestore.ArrayRemove(dataSubjectId),
		})
		mongoUpdates = append(mongoUpdates, bson.D{{Key: "$pull", Value: bson.D{{Key: "users", Value: dataSubjectId}}}})
	}

	fieldsToUpdate = pal.FieldUpdates{
		FirestoreUpdates: firestoreUpdates,
		MongoUpdates:     mongoUpdates,
	}

	return
}

func handleDeletionMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) (nodesToTraverse []pal.Locator, deleteNode bool, fieldsToUpdate pal.FieldUpdates) {
	deleteNode = true
	return
}

func handleDeletionDirectMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) (nodesToTraverse []pal.Locator, deleteNode bool, fieldsToUpdate pal.FieldUpdates) {
	nodesToTraverse = append(nodesToTraverse, pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Filters: []pal.Filter{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
		MongoLocator: pal.MongoLocator{
			Collection: "messages",
			Filter:     bson.D{{Key: "userId", Value: dataSubjectId}, {Key: "chatId", Value: dbObj["_id"]}},
		},
	})

	deleteNode = false

	fieldsToUpdate = pal.FieldUpdates{
		MongoUpdates: []interface{}{
			bson.D{{Key: "$pull", Value: bson.D{{Key: "$pull", Value: bson.D{{Key: "messages", Value: bson.D{{Key: "userId", Value: dataSubjectId}, {Key: "chatId", Value: dbObj["_id"]}}}}}}}},
		},
	}

	return
}
